src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/JsonValueConverterKeyValuePair.cs

```diff
@@ -62,15 +68,17 @@ internal sealed class JsonKeyValuePairConverter<TKey, TValue> : JsonConverter<Ke
             }
 
             propertyName = reader.GetString()!;
-            if (propertyName == ValueName)
+            if (propertyName == KeyName)
             {
-                v = ReadProperty<TValue>(ref reader, typeToConvert, options);
-                valueSet = true;
+                reader.Read();
+                k = JsonSerializer.Deserialize<TKey>(ref reader, options, ref state, KeyName);
```

Comment:
No it is slower mostly since a dictionary lookup needs to occur to get the converter but also because there is a small amount of overhead -- extra functions\validation and boxing+unboxing for structs (which we should fix after this PR).

If you know the converter or can cache it, then that should be used. There's a "todo" in this file to cache the converter. For built-in converters, caching the converter is easy since the cached converter(s) are determined by using the `JsonSerializerOptions` instance and the converter instance is re-created for every `JsonSerializerOptions` instance so we don't have to worry about different results.

Reply:
Calling the Serializer directly is almost as fast -- there is a little extra functions\validation and boxing+unboxing for structs (which we should fix after this PR). In general, we should recommend just re-entering the serializer since the overhead will be very small.

However, if you know the converter or can cache it, then that should be used. There's a "todo" in this file to cache the converter. For built-in converters like `KeyValueValue`, caching the forwarding converters (for `TKey`, `TValue`) is easy since the cached converter(s) are determined by using the `JsonSerializerOptions` instance and the converter instance is re-created for every `JsonSerializerOptions` instance so we don't have to worry about different results.